### PR TITLE
internal/envoy: configurable access log format

### LINKED ISSUE #3576: Add option to remove query strings from access log
**Feature request**

Envoy will log the complete request path, including the query string like `?supersecret=password` in this example

```
[2021-04-14T16:36:00.361Z] "GET /foo?supersecret=password HTTP/1.1" 200 - 0 463 6 3 "-" "HTTPie/1.0.3" "837aa8dc-344f-4faa-b7d5-c9cce1028519" "localhost:8080" "127.0.0.1:8081"
```

It would be great to provide an option to remove query string from the access log, since it may contain sensitive information introducing a potential [vulnerability](https://owasp.org/www-community/vulnerabilities/Information_exposure_through_query_strings_in_url). If the problem cannot be fixed in the proxied application itself, the vulnerability could be mitigated by sanitazing the access logs.

**Potential options**

1) Contour could add HTTP filter called [Header-To-Metadata](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/header_to_metadata/v3/header_to_metadata.proto) and use it to remove query string from :PATH header and set the result in dynamic metadata, see https://github.com/envoyproxy/envoy/pull/15711#issuecomment-810511085.  Then Contour could provide the user an option to change the [default access log format string](https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/access_log/usage#default-format-string), where they would replace `%REQ(X-ENVOY-ORIGINAL-PATH?:PATH)%` (including the query string) with `%DYNAMIC_METADATA(projectcontour.io:path)%` (query string sanitized)

2) Proceed with the PR https://github.com/envoyproxy/envoy/pull/15711 (or something similar) which intends to introduce a new extension command e.g. `REQ_WITHOUT_QUERY` for the purpose of sanitizing the headers. Next step would be for Contour to enable this extension in the access log configuration within HTTP Connection Manager. Then provide the user an option to change the default access log format string where they would replace `%REQ(X-ENVOY-ORIGINAL-PATH?:PATH)%` with `%REQ_WITHOUT_QUERY(X-ENVOY-ORIGINAL-PATH?:PATH)%`

Option (1) is possible to implement with existing versions of Envoy, while (2) would introduce a version dependency.

### PR DESCRIPTION
**Pull request by tsaarni**
This change adds following features:

* New configuration option that allows setting the format string for Envoy's text based access log.
* Adds support for Envoy's REQ_WITHOUT_QUERY command operator, which can be used to avoid writing query string to access log.

Fixes #3576

Signed-off-by: Tero Saarni <tero.saarni@est.tech>

### CODE REVIEW

**Comment by tsaarni** [comment]
xref #3576

**Comment by skriss** [comment]
@tsaarni sorry we've been slow to provide feedback here. In principle I think this feature makes sense. With regard to the command operators, I'm not sure if you've seen but we've accepted a contribution to do something similar for headers: https://projectcontour.io/docs/v1.16.0/config/request-rewriting/#dynamic-header-values and https://github.com/projectcontour/contour/blob/main/internal/dag/policy.go#L262-L305. With that feature, we included an allow-list of possible command operators in the Contour code. That makes me wonder, do you know what Envoy's behavior is if a malformed access log format string is passed? Will it NACK the config update, or ignore invalid command operators, or something else? That may inform how paranoid we need to be about ensuring the passed value is valid.

**Comment by tsaarni** [comment]
Thanks @skriss for responding!

> With that feature, we included an allow-list of possible command operators in the Contour code. 

Yes, I think I would need to do that anyways: my underlying goal https://github.com/projectcontour/contour/issues/3576 would require use of Envoy's (forthcoming) log format command operator extension. These formatter extension needs to be enabled explicitly before use.  For Contour to allow using such command operator, it would need to recognize the command enable the corresponding extension like this https://github.com/projectcontour/contour/issues/3576#issuecomment-853590858.  On the other hand, if the command is not used, the extension would not need to be enabled (or it could be always enabled, but this would mean Envoy needs to be always compiled with that extension).

The JSON access log validation does bit more: it checks the [command operator syntax with regex]( https://github.com/projectcontour/contour/blob/05d6ee3f557ada10364f01fced50b2aecbae6bf1/pkg/config/parameters.go#L123) but I'm bit uncertain how far to go? It can only capture some specific errors, e.g. it cannot validate if header name in `%REQ(MY-HTTP-HEADER)%` is correct. 

If letting Envoy to do the validation, one gets quite good error message in NACK, at least in this case. I guess the biggest difference is that when Contour is validating, it will `exit()` and crash loop. When Envoy validates, it will be just error printouts in logs.  In both cases it is the same from traffic perspective: Envoy will serve requests with old config (as long as it happens to restart).

If new command operators are introduced, maintaining allow-list will always mean code impact on Contour. Though it seems that Envoy encourages to implement new command operators as formatter extensions, so it would mean code impact anyways - for maintaining a list of extensions to enable.

> That makes me wonder, do you know what Envoy's behavior is if a malformed access log format string is passed? Will it NACK the config update, or ignore invalid command operators, or something else? That may inform how paranoid we need to be about ensuring the passed value is valid.

It will NACK the config update, and give error message which will be visible both in Contour and Envoy logs, example here https://github.com/projectcontour/contour/issues/3576#issuecomment-852781960.

**Comment by tsaarni** [comment]
I've now adapted the existing validation code from JSON access log fields so that the same validation is also done for text access log format configuration.

Meanwhile, the access log formatter extension which I was working in Envoy was merged https://github.com/envoyproxy/envoy/pull/15711.  I uplifted `go-control-plane` to a **unreleased version** to experiment with this in Contour.  The proposed implementation now recognizes when user has `REQ_WITHOUT_QUERY` command operator in either JSON or text based access log format string, and enables the formatter extension.  When the command is not present the formatter is not enabled.  This should allow user to run old Envoy version or Envoy version with disabled extension.

File: pkg/config/parameters.go
Comment (skriss) [inline pkg/config/parameters.go:254]: Just wondering if there's a way to make it more obvious for the next developer that if they update this section for a new extension, they also need to update `internal/envoy/v3/accesslog.go` - at a minimum let's just add a comment, I think that should be sufficient.

File: internal/envoy/v3/accesslog.go
Comment (skriss) [inline internal/envoy/v3/accesslog.go:102]: a comment here cross-referencing `AccessLogFormatterExtensions()` in `pkg/config/parameters.go` would be helpful.

**Review by skriss — COMMENTED** [review]
Thanks for the PR @tsaarni - I think this looks pretty reasonable. I had some minor inline comments. I still need to look at the tests in some more detail as well.

It'd also be nice to find some place to document this more thoroughly than the config reference docs. One option would be to expand the JSON access logging guide (https://projectcontour.io/guides/structured-logs/) to cover non-JSON access logging as well, and to include the new operator / extension there. We could do that in a separate PR that we don't merge until the next release, since the guides aren't versioned so any changes are "live" immediately.

**Review by stevesloka — COMMENTED** [review]
I need to do some real tests with this. Would be super nice to have some E2E tests or follow up with some in another PR. The only downside is we need to test against an unreleased version of Envoy at the moment.

File: site/content/docs/main/configuration.md
Comment (tsaarni) [reply site/content/docs/main/configuration.md]: I added note about this.  While at it, I also added similar note to `json-fields` which I believe has effect only when `accesslog-format` is `json`.

File: pkg/config/parameters.go
Comment (tsaarni) [reply pkg/config/parameters.go]: Fixed.

File: internal/envoy/v3/accesslog.go
Comment (tsaarni) [reply internal/envoy/v3/accesslog.go]: Fixed.

File: pkg/config/parameters.go
Comment (tsaarni) [reply pkg/config/parameters.go:255]: I've now added notes / cross reference to both functions: `AccessLogFormatterExtensions()` in `parameters.go` and `extensionConfig()` in `accesslog.go` to remind that these should be updated as a pair.

File: internal/envoy/v3/accesslog.go
Comment (tsaarni) [reply internal/envoy/v3/accesslog.go:38]: Yes, the same parser is now used for both JSON and text based logs. If for some reason invalid format would still get through parser, I've tested how Envoy would react, for results see https://github.com/projectcontour/contour/issues/3576#issuecomment-852781960.

There is also one very common error that could be validated: there needs to be `\n` at the end of the line.
When it is not included, logs are not flushed until certain amount of logs have been written.  And obviously, everything comes as one long line. I think I saw old discussions about automatically adding this, but it is still not done today.

File: examples/contour/01-contour-config.yaml
Comment (tsaarni) [reply examples/contour/01-contour-config.yaml:57]: I've shortened the text by removing the link to Envoy docs and the full example (still leaving short example value `accesslog-format-string: "...\n"`)

The config file is bit hard to read in general.  Maybe one small improvement would be changing indentation of fields-vs-comments, see e.g. [sshd_config](https://raw.githubusercontent.com/openssh/openssh-portable/master/sshd_config)

```yaml
# first comment
#first-field:
# second comment
#second-field:
```

Comment (tsaarni) [reply examples/contour/01-contour-config.yaml:57]: d'oh, it obviously fails anyways to be readable due to structs having indentations :-D

```yaml
# first comment
#first-field:
#  sub-field:
# second comment
```

maybe double-hash for comments:

```yaml
## first comment
#first-field:
#  sub-field:
## second comment
```

File: internal/envoy/v3/accesslog.go
Comment (youngnick) [reply internal/envoy/v3/accesslog.go:38]: Oof, yeah, that `\n` one is a great candidate for us to check, as you probably would never actually want that behavior.

Comment (skriss) [reply internal/envoy/v3/accesslog.go:38]: Yeah, I think I'm a 👍 to enforcing/adding the `\n`.

File: go.mod
Comment (skriss) [reply go.mod:8]: I think as long as tests are passing (which they are currently), I'm OK with shipping with a specific commit here.

File: examples/contour/01-contour-config.yaml
Comment (skriss) [reply examples/contour/01-contour-config.yaml:57]: Yeah, this formatting is not great. I filed https://github.com/projectcontour/contour/issues/3843 to make some improvements here.

File: internal/envoy/v3/accesslog.go
Comment (tsaarni) [reply internal/envoy/v3/accesslog.go:38]: I've now added check for `\n`.

**Comment by tsaarni** [comment]
I believe I have covered current comments for now. Please take a look.

I'll create another issue for updating https://projectcontour.io/guides/structured-logs/ (or some other page) with detailed usage docs.

One question regarding API:
The new parameter `accesslog-format-string` will be effective when `accesslog-format: envoy` is set. Maybe value `envoy` was originally intended to refer to *"the format that Envoy defines"* and that is why it is called like that.  In theory `text` would make more sense now, as an alternative to `json`, but is it OK to keep using `envoy`?  

I guess adding third option (`envoy` = default Envoy format, `json` = custom JSON logs, `text` =custom text logs) would be equally confusing as just continuing to use `envoy` and `json`, former meaning any text based log, default or custom.

**Comment by skriss** [comment]
> One question regarding API:
> The new parameter `accesslog-format-string` will be effective when `accesslog-format: envoy` is set. Maybe value `envoy` was originally intended to refer to _"the format that Envoy defines"_ and that is why it is called like that. In theory `text` would make more sense now, as an alternative to `json`, but is it OK to keep using `envoy`?
> 
> I guess adding third option (`envoy` = default Envoy format, `json` = custom JSON logs, `text` =custom text logs) would be equally confusing as just continuing to use `envoy` and `json`, former meaning any text based log, default or custom.

I think leaving it as `envoy` is OK - to me it just means the non-JSON format, the exact fields of which can now be customized.

**Review by skriss — APPROVED** [review]
no further comments from me, LGTM. thanks @tsaarni!

**Review by youngnick — APPROVED** [review]
This LGTM, but before we merge, I just need to check: @tsaarni, what version of Envoy is the required extension available in? Do we need to hold this PR until after v1.17.0 to wait for the new Envoy version?

If it works with v1.18.3, then we're good to go, if not, we should hold off merging, since 1.17.0 is due out tomorrow (my time).

**Comment by tsaarni** [comment]
> If it works with v1.18.3, then we're good to go, if not, we should hold off merging, since 1.17.0 is due out tomorrow (my time).

Otherwise it will work **but** if user adds the new command operator extension `REQ_WITHOUT_QUERY` in a custom format, then Contour will try to enable the Envoy formatter extension and fail until there is Envoy 1.19.0.

**Comment by youngnick** [comment]
Hmm, okay, thanks. While I know it's the reason you're making the change, I don't think that we can introduce this feature with a known-not-working field in it, given the current level of Envoy support.

In that case, we have two options:
- Hold this PR to 1.18, which should be released around the same time Envoy 1.19 is
- Remove the bits to do REQ_WITHOUT_QUERY for now, and put them back in once Envoy 1.19 is available.

@tsaarni, what do you think?

**Comment by tsaarni** [comment]
> * Hold this PR to 1.18, which should be released around the same time Envoy 1.19 is

Yes, lets wait for the Envoy release first.   It will avoid some extra work.

**Comment by sunjayBhatia** [comment]
I'll add the do not merge tag for now so we dont accidentally merge this

**Comment by skriss** [comment]
Just merged https://github.com/projectcontour/contour/pull/3887, so I *think* this is ready to go now, just need a quick rebase to fix merge conflicts.

**Comment by tsaarni** [comment]
I've now rebased and run manual test with the current Envoy version.

I did not look yet how this could be fully automated in the integration tests but I've written manual test procedure here https://gist.github.com/tsaarni/6e8be52179d775bf2a90b6e009e5ef94

**Comment by skriss** [comment]
This is ready to merge from my perspective now that we've updated Envoy, cc @youngnick to confirm

**Comment by skriss** [comment]
Sorry @tsaarni, needs another rebase.

**Comment by tsaarni** [comment]
@skriss no problem, rebase done!

**Comment by skriss** [comment]
Awesome, thanks. I''ll merge this first thing next week unless I hear otherwise from @youngnick

**Review by youngnick — APPROVED** [review]
Envoy 1.19 is out now, so I thnk we are good to go.

**Pull request merged**
